### PR TITLE
Fixed: Show new error message when output schema is not maching input…

### PR DESCRIPTION
**Pull request by vjosadika**
… schema
Jira ticket: https://issues.cask.co/browse/PLUGIN-132

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -22,14 +22,14 @@
 
   <groupId>io.cdap.plugin</groupId>
   <artifactId>dynamic-spark</artifactId>
-  <version>2.3.0-SNAPSHOT</version>
+  <version>2.3.0</version>
```
Comment (CuriousVini) [inline pom.xml]: develop branch should always use snapshot versions

File: pom.xml [hunk 2]
```diff
@@ -22,14 +22,14 @@
...
 
   <properties>
     <!-- properties for script build step that creates the config files for the artifacts -->
     <data.pipeline.parent>system:cdap-data-pipeline[6.0.0-SNAPSHOT,7.0.0-SNAPSHOT)</data.pipeline.parent>
     <data.stream.parent>system:cdap-data-streams[6.0.0-SNAPSHOT,7.0.0-SNAPSHOT)</data.stream.parent>
 
-    <cdap.version>6.2.0-SNAPSHOT</cdap.version>
```
Comment (CuriousVini) [inline pom.xml:32]: this change is not needed

File: src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java [hunk 3]
```diff
@@ -110,6 +110,11 @@ public void prepareRun(SparkPluginContext context) throws Exception {
       // If there is no output schema configured, derive it from the DataFrame
       // Otherwise, assume the DataFrame has the correct schema already
       outputSchema = DataFrames.toSchema((DataType) invokeDataFrameMethod(result, "schema"));
+    }else{
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java]: space after }

File: src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java [hunk 4]
```diff
@@ -110,6 +110,11 @@ public void prepareRun(SparkPluginContext context) throws Exception {
...
+      Schema dataSchema = DataFrames.toSchema((DataType) invokeDataFrameMethod(result, "schema"));
+      if(dataSchema.getFields().size() < outputSchema.getFields().size()){
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java]: space after `if`

Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java]: space before `{`

File: src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java [hunk 5]
```diff
@@ -110,6 +110,11 @@ public void prepareRun(SparkPluginContext context) throws Exception {
...
+        throw new Exception("Invalid schema. Output schema is not matching input schema.");
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java]: if this check is happening at validation time, why are we not using failure collector?

**Review by CuriousVini — COMMENTED** [review]
Please make sure plugin code is formatted as per checkstyle requirements

File: src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java (hunk 5 above)
Comment (vjosadika) [reply src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java]: Fixed

File: src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java [hunk 6]
```diff
@@ -110,6 +111,13 @@ public void prepareRun(SparkPluginContext context) throws Exception {
       // If there is no output schema configured, derive it from the DataFrame
       // Otherwise, assume the DataFrame has the correct schema already
       outputSchema = DataFrames.toSchema((DataType) invokeDataFrameMethod(result, "schema"));
+    } else {
+      Schema dataSchema = DataFrames.toSchema((DataType) invokeDataFrameMethod(result, "schema"));
+      if (dataSchema.getFields().size() < outputSchema.getFields().size()) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java]: This is just based on size. Why are we not checking types ?

File: src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java [hunk 7]
```diff
@@ -110,6 +111,13 @@ public void prepareRun(SparkPluginContext context) throws Exception {
...
+        FailureCollector collector = context.getFailureCollector();
+        collector.addFailure("Invalid schema.","Output schema is not matching input schema.");
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java]: The schema is not invalid. so the error message should be `Schema mismatch.`

**Review by CuriousVini — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java [hunk 8]
```diff
@@ -110,6 +111,13 @@ public void prepareRun(SparkPluginContext context) throws Exception {
       // If there is no output schema configured, derive it from the DataFrame
       // Otherwise, assume the DataFrame has the correct schema already
       outputSchema = DataFrames.toSchema((DataType) invokeDataFrameMethod(result, "schema"));
+    } else {
+      Schema dataSchema = DataFrames.toSchema((DataType) invokeDataFrameMethod(result, "schema"));
+      if (!dataSchema.isCompatible(outputSchema)) {
+        FailureCollector collector = context.getFailureCollector();
+        collector.addFailure("Schema mismatch.", "Output schema is not matching input schema.");
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/spark/dynamic/ScalaSparkCompute.java:118]: Could you please rephrase:
`"Output schema is not compatible with input schema.", null`

**Review by CuriousVini — APPROVED** [review]
Please fix and squash commits

**Pull request merged**